# Question 2: If Quadeca’s music did change, did the success of *Insecure* influence Quadeca to make music that reflected mainstream musical themes and styles?

---

## Setup

In [1]:
from collections import Counter

In [87]:
# import notebooks to run
%run /Commjhub/jupyterhub/comm318_fall2019/raymon-shi/comm313_S21/comm313_S21_Final_Project/data/functions.ipynb

In [3]:
# important global variables
%store -r a_aone_tokens
%store -r a_atwo_tokens
%store -r after_insecure_final
%store -r chart_2018_final
%store -r chart_2019_final
%store -r chart_2020_final

characters_to_strip = '?().[]!,"-’' + "/'"

---

## Word Frequency and N-Gram Frequency Lists

For the word frequency and n-gram frequency lists, I will create the n-gram counters like I would normally do, but this time, I will also get the intersection between the most commonly repeated words and phrases and the most commonly repeated words and phrases between all songs. I believe that getting this intersection will help get the best of both worlds. 

### Combined Albums After the Insecure Diss Track

In [4]:
# building the counters
after_all_lyrics = list()

for song_dict in after_insecure_final[0]:
    song_lyrics = song_dict['lyrics']
    after_all_lyrics.append(song_lyrics)
    
for song_dict in after_insecure_final[1]:
    song_lyrics = song_dict['lyrics']
    after_all_lyrics.append(song_lyrics)

In [36]:
after_common_ctr = Counter()
after_song_freq_ctr = Counter()
after_cleaned_all = get_cleaned_album_all(after_all_lyrics, after_common_ctr, after_song_freq_ctr, 1)

In [6]:
# most common 1-grams with stopwords
after_common_ctr.most_common(50)

[('i', 792),
 ('the', 498),
 ('a', 337),
 ('yeah', 320),
 ('you', 317),
 ('to', 283),
 ('my', 270),
 ('im', 233),
 ('me', 231),
 ('in', 216),
 ('they', 206),
 ('and', 203),
 ('it', 197),
 ('like', 178),
 ('huh', 157),
 ('that', 151),
 ('on', 142),
 ('uh', 140),
 ('but', 131),
 ('all', 121),
 ('just', 108),
 ('up', 107),
 ('of', 107),
 ('be', 99),
 ('its', 99),
 ('this', 96),
 ('for', 96),
 ('no', 87),
 ('when', 87),
 ('dont', 86),
 ('so', 84),
 ('what', 79),
 ('if', 75),
 ('been', 73),
 ('shit', 72),
 ('got', 71),
 ('aint', 69),
 ('your', 64),
 ('with', 63),
 ('know', 62),
 ('thats', 62),
 ('was', 62),
 ('not', 61),
 ('is', 60),
 ('now', 60),
 ('never', 57),
 ('cant', 57),
 ('at', 56),
 ('do', 54),
 ('think', 54)]

In [7]:
# most popular across multiple songs 1-grams with stopwords
after_song_freq_ctr.most_common(50)

[('the', 30),
 ('i', 30),
 ('to', 30),
 ('and', 30),
 ('you', 29),
 ('it', 29),
 ('a', 29),
 ('of', 28),
 ('me', 28),
 ('all', 27),
 ('in', 27),
 ('my', 26),
 ('just', 26),
 ('im', 26),
 ('yeah', 26),
 ('on', 26),
 ('when', 26),
 ('that', 25),
 ('be', 25),
 ('but', 25),
 ('for', 25),
 ('if', 24),
 ('its', 24),
 ('been', 24),
 ('never', 24),
 ('like', 24),
 ('they', 23),
 ('with', 23),
 ('so', 23),
 ('this', 23),
 ('up', 22),
 ('dont', 22),
 ('your', 22),
 ('no', 22),
 ('got', 21),
 ('at', 21),
 ('know', 21),
 ('uh', 20),
 ('now', 20),
 ('thats', 20),
 ('shit', 20),
 ('do', 20),
 ('was', 20),
 ('not', 20),
 ('even', 19),
 ('say', 19),
 ('want', 19),
 ('cant', 19),
 ('get', 19),
 ('still', 19)]

In [8]:
# finding the intersection between the most commonly repeated terms and the most commonly popular terms across multiple songs
inter_ctr_gram_1 = after_common_ctr & after_song_freq_ctr
inter_ctr_gram_1.most_common(50)

[('i', 30),
 ('the', 30),
 ('to', 30),
 ('and', 30),
 ('a', 29),
 ('you', 29),
 ('it', 29),
 ('of', 28),
 ('me', 28),
 ('in', 27),
 ('all', 27),
 ('my', 26),
 ('im', 26),
 ('just', 26),
 ('yeah', 26),
 ('on', 26),
 ('when', 26),
 ('be', 25),
 ('that', 25),
 ('but', 25),
 ('for', 25),
 ('if', 24),
 ('like', 24),
 ('been', 24),
 ('its', 24),
 ('never', 24),
 ('this', 23),
 ('they', 23),
 ('so', 23),
 ('with', 23),
 ('no', 22),
 ('up', 22),
 ('your', 22),
 ('dont', 22),
 ('know', 21),
 ('got', 21),
 ('at', 21),
 ('thats', 20),
 ('do', 20),
 ('shit', 20),
 ('uh', 20),
 ('not', 20),
 ('now', 20),
 ('was', 20),
 ('even', 19),
 ('say', 19),
 ('want', 19),
 ('cant', 19),
 ('get', 19),
 ('still', 19)]

In [9]:
after_common_ctr_2 = Counter()
after_song_freq_ctr_2 = Counter()
after_cleaned_all_2 = get_cleaned_album_all(after_all_lyrics, after_common_ctr_2, after_song_freq_ctr_2, 2)

In [10]:
after_common_ctr_2.most_common(50)

[('in the', 64),
 ('on my', 54),
 ('yeah i', 45),
 ('yeah yeah', 45),
 ('i just', 44),
 ('in my', 43),
 ('but i', 38),
 ('i got', 33),
 ('i dont', 33),
 ('i was', 31),
 ('and i', 31),
 ('when i', 29),
 ('if you', 28),
 ('that i', 27),
 ('huh i', 27),
 ('like a', 27),
 ('i think', 25),
 ('what i', 25),
 ('i cant', 25),
 ('work until', 25),
 ('on the', 24),
 ('to the', 23),
 ('i been', 23),
 ('i had', 22),
 ('in a', 22),
 ('uh yeah', 22),
 ('to me', 21),
 ('ive been', 21),
 ('this shit', 20),
 ('had to', 20),
 ('think i', 19),
 ('they like', 19),
 ('i need', 19),
 ('of the', 19),
 ('huh yeah', 19),
 ('you know', 19),
 ('i know', 18),
 ('got a', 18),
 ('im a', 18),
 ('ooh ooh', 18),
 ('im not', 18),
 ('the same', 18),
 ('be the', 17),
 ('for real', 17),
 ('all of', 17),
 ('im the', 17),
 ('so i', 17),
 ('i say', 17),
 ('until it', 17),
 ('yeah huh', 16)]

In [11]:
after_song_freq_ctr_2.most_common(50)

[('in the', 25),
 ('in my', 18),
 ('i just', 17),
 ('i got', 17),
 ('on my', 17),
 ('yeah i', 15),
 ('but i', 15),
 ('on the', 15),
 ('when i', 15),
 ('that i', 14),
 ('of the', 14),
 ('i cant', 13),
 ('i dont', 13),
 ('im a', 13),
 ('i think', 13),
 ('yeah yeah', 12),
 ('i been', 12),
 ('feel like', 12),
 ('i was', 12),
 ('in a', 12),
 ('me i', 12),
 ('to the', 12),
 ('so i', 12),
 ('what i', 11),
 ('i need', 11),
 ('if you', 11),
 ('to me', 11),
 ('you know', 11),
 ('if i', 10),
 ('for real', 10),
 ('this shit', 10),
 ('got a', 10),
 ('think i', 10),
 ('like a', 10),
 ('i know', 10),
 ('all of', 10),
 ('and i', 10),
 ('when im', 10),
 ('that shit', 10),
 ('for me', 10),
 ('ive been', 10),
 ('be the', 9),
 ('yeah but', 9),
 ('to be', 9),
 ('like im', 9),
 ('i do', 9),
 ('all the', 9),
 ('uh yeah', 9),
 ('i am', 9),
 ('but you', 9)]

In [12]:
inter_ctr_gram_2 = after_common_ctr_2 & after_song_freq_ctr_2
inter_ctr_gram_2.most_common(50)

[('in the', 25),
 ('in my', 18),
 ('i got', 17),
 ('on my', 17),
 ('i just', 17),
 ('but i', 15),
 ('yeah i', 15),
 ('when i', 15),
 ('on the', 15),
 ('that i', 14),
 ('of the', 14),
 ('i think', 13),
 ('i dont', 13),
 ('im a', 13),
 ('i cant', 13),
 ('in a', 12),
 ('to the', 12),
 ('me i', 12),
 ('i was', 12),
 ('feel like', 12),
 ('yeah yeah', 12),
 ('i been', 12),
 ('so i', 12),
 ('what i', 11),
 ('i need', 11),
 ('if you', 11),
 ('to me', 11),
 ('you know', 11),
 ('i know', 10),
 ('got a', 10),
 ('if i', 10),
 ('this shit', 10),
 ('for real', 10),
 ('think i', 10),
 ('like a', 10),
 ('all of', 10),
 ('and i', 10),
 ('that shit', 10),
 ('when im', 10),
 ('for me', 10),
 ('ive been', 10),
 ('i do', 9),
 ('be the', 9),
 ('all the', 9),
 ('like im', 9),
 ('uh yeah', 9),
 ('yeah but', 9),
 ('to be', 9),
 ('but you', 9),
 ('i am', 9)]

In [13]:
after_common_ctr_3 = Counter()
after_song_freq_ctr_3 = Counter()
after_cleaned_all_3 = get_cleaned_album_all(after_all_lyrics, after_common_ctr_3, after_song_freq_ctr_3, 3)

In [14]:
after_common_ctr_3.most_common(50)

[('yeah yeah yeah', 21),
 ('work until it', 17),
 ('aint nothing usual', 16),
 ('i had to', 14),
 ('until it work', 13),
 ('i got a', 12),
 ('i woke up', 12),
 ('gon put in', 12),
 ('put in work', 12),
 ('in work until', 12),
 ('for a new', 11),
 ('i think i', 10),
 ('woke up this', 10),
 ('im on my', 9),
 ('i need to', 9),
 ('usual bout me', 9),
 ('we gon put', 9),
 ('it work until', 9),
 ('i just wanna', 8),
 ('for real i', 8),
 ('i love you', 8),
 ('looking for a', 8),
 ('up in the', 8),
 ('the reason that', 8),
 ('nothing usual bout', 8),
 ('shades of us', 8),
 ('i just pulled', 7),
 ('just pulled up', 7),
 ('if you gon', 7),
 ('i dont know', 7),
 ('ooh ooh ooh', 7),
 ('you and me', 7),
 ('im the only', 7),
 ('the only one', 7),
 ('look at my', 7),
 ('they do not', 7),
 ('reason that they', 7),
 ('i used to', 7),
 ('the cameras on', 7),
 ('no matter how', 7),
 ('at the ground', 7),
 ('i do not', 6),
 ('on my own', 6),
 ('at my fit', 6),
 ('life like a', 6),
 ('that they never', 6),

In [15]:
after_song_freq_ctr_3.most_common(50)

[('i got a', 7),
 ('i think i', 7),
 ('yeah yeah yeah', 7),
 ('i need to', 6),
 ('i just wanna', 5),
 ('for real i', 5),
 ('i do not', 5),
 ('i love you', 5),
 ('up in the', 5),
 ('i had to', 4),
 ('im on my', 4),
 ('think i need', 4),
 ('yeah but i', 4),
 ('uh yeah i', 4),
 ('all of the', 4),
 ('i am not', 4),
 ('i dont know', 4),
 ('how i feel', 4),
 ('all of my', 4),
 ('i feel like', 4),
 ('in my mind', 4),
 ('be the same', 4),
 ('no matter how', 4),
 ('i know i', 3),
 ('i wish i', 3),
 ('just pulled up', 3),
 ('i just pulled', 3),
 ('i dont wanna', 3),
 ('feel like im', 3),
 ('wanna be the', 3),
 ('i was in', 3),
 ('me but i', 3),
 ('i was too', 3),
 ('i had a', 3),
 ('in the wrong', 3),
 ('huh i got', 3),
 ('they say im', 3),
 ('i did it', 3),
 ('huh yeah huh', 3),
 ('i dont want', 3),
 ('the back of', 3),
 ('i just want', 3),
 ('when i go', 3),
 ('it all in', 3),
 ('in the back', 3),
 ('thats why i', 3),
 ('they all wanna', 3),
 ('yeah yeah i', 3),
 ('but i cant', 3),
 ('the only

In [16]:
inter_ctr_gram_3 = after_common_ctr_3 & after_song_freq_ctr_3
inter_ctr_gram_3.most_common(50)

[('i got a', 7),
 ('i think i', 7),
 ('yeah yeah yeah', 7),
 ('i need to', 6),
 ('i just wanna', 5),
 ('for real i', 5),
 ('i do not', 5),
 ('i love you', 5),
 ('up in the', 5),
 ('im on my', 4),
 ('i had to', 4),
 ('uh yeah i', 4),
 ('yeah but i', 4),
 ('think i need', 4),
 ('all of the', 4),
 ('i am not', 4),
 ('i dont know', 4),
 ('how i feel', 4),
 ('all of my', 4),
 ('i feel like', 4),
 ('in my mind', 4),
 ('be the same', 4),
 ('no matter how', 4),
 ('i know i', 3),
 ('wanna be the', 3),
 ('i just pulled', 3),
 ('just pulled up', 3),
 ('i wish i', 3),
 ('feel like im', 3),
 ('i dont wanna', 3),
 ('i was too', 3),
 ('i had a', 3),
 ('i did it', 3),
 ('they say im', 3),
 ('me but i', 3),
 ('i was in', 3),
 ('in the wrong', 3),
 ('huh i got', 3),
 ('huh yeah huh', 3),
 ('when i go', 3),
 ('it all in', 3),
 ('in the back', 3),
 ('the back of', 3),
 ('i dont want', 3),
 ('i just want', 3),
 ('yeah yeah i', 3),
 ('they all wanna', 3),
 ('thats why i', 3),
 ('but i cant', 3),
 ('fuck all

### Combined Billboard Rap Charts from 2018, 2019, and 2020

In [17]:
all_rap_chart_lyrics = list()

for song_dict in chart_2018_final:
    song_lyrics = song_dict['lyrics']
    all_rap_chart_lyrics.append(song_lyrics)
    
for song_dict in chart_2019_final:
    song_lyrics = song_dict['lyrics']
    all_rap_chart_lyrics.append(song_lyrics)
    
for song_dict in chart_2020_final:
    song_lyrics = song_dict['lyrics']
    all_rap_chart_lyrics.append(song_lyrics)
    
len(all_rap_chart_lyrics)

150

In [18]:
rap_chart_common_ctr = Counter()
rap_chart_song_freq_ctr = Counter()
rap_chart_cleaned_all = get_cleaned_album_all(all_rap_chart_lyrics, rap_chart_common_ctr, rap_chart_song_freq_ctr, 1)

In [19]:
rap_chart_common_ctr.most_common(50)

[('i', 3663),
 ('the', 2919),
 ('you', 2335),
 ('a', 2085),
 ('it', 1473),
 ('my', 1451),
 ('me', 1370),
 ('yeah', 1336),
 ('and', 1184),
 ('on', 1163),
 ('to', 1143),
 ('in', 1129),
 ('im', 1108),
 ('like', 988),
 ('that', 936),
 ('with', 794),
 ('got', 733),
 ('up', 727),
 ('she', 617),
 ('dont', 558),
 ('for', 547),
 ('bitch', 538),
 ('your', 521),
 ('we', 501),
 ('get', 499),
 ('they', 489),
 ('know', 484),
 ('this', 464),
 ('its', 437),
 ('aint', 436),
 ('no', 434),
 ('of', 434),
 ('all', 430),
 ('be', 424),
 ('ayy', 417),
 ('just', 407),
 ('go', 402),
 ('nigga', 401),
 ('but', 394),
 ('fuck', 372),
 ('shit', 368),
 ('out', 365),
 ('her', 361),
 ('down', 342),
 ('what', 307),
 ('is', 303),
 ('now', 278),
 ('so', 278),
 ('when', 277),
 ('cant', 265)]

In [20]:
rap_chart_song_freq_ctr.most_common(50)

[('i', 150),
 ('the', 149),
 ('a', 148),
 ('to', 148),
 ('my', 147),
 ('you', 147),
 ('me', 147),
 ('im', 144),
 ('and', 144),
 ('on', 142),
 ('in', 141),
 ('it', 137),
 ('that', 136),
 ('dont', 133),
 ('up', 132),
 ('got', 130),
 ('like', 130),
 ('with', 129),
 ('yeah', 129),
 ('know', 118),
 ('no', 117),
 ('be', 116),
 ('all', 115),
 ('get', 115),
 ('this', 114),
 ('but', 114),
 ('out', 113),
 ('they', 112),
 ('of', 112),
 ('your', 109),
 ('just', 109),
 ('its', 108),
 ('we', 108),
 ('for', 106),
 ('she', 104),
 ('when', 104),
 ('shit', 103),
 ('so', 102),
 ('aint', 102),
 ('go', 97),
 ('fuck', 94),
 ('back', 92),
 ('is', 91),
 ('bitch', 90),
 ('if', 89),
 ('from', 89),
 ('make', 87),
 ('her', 87),
 ('do', 85),
 ('was', 85)]

In [21]:
inter_rap_gram_1 = rap_chart_common_ctr & rap_chart_song_freq_ctr
inter_rap_gram_1.most_common(50)

[('i', 150),
 ('the', 149),
 ('a', 148),
 ('to', 148),
 ('me', 147),
 ('you', 147),
 ('my', 147),
 ('and', 144),
 ('im', 144),
 ('on', 142),
 ('in', 141),
 ('it', 137),
 ('that', 136),
 ('dont', 133),
 ('up', 132),
 ('like', 130),
 ('got', 130),
 ('yeah', 129),
 ('with', 129),
 ('know', 118),
 ('no', 117),
 ('be', 116),
 ('all', 115),
 ('get', 115),
 ('this', 114),
 ('but', 114),
 ('out', 113),
 ('they', 112),
 ('of', 112),
 ('your', 109),
 ('just', 109),
 ('its', 108),
 ('we', 108),
 ('for', 106),
 ('when', 104),
 ('she', 104),
 ('shit', 103),
 ('so', 102),
 ('aint', 102),
 ('go', 97),
 ('fuck', 94),
 ('back', 92),
 ('is', 91),
 ('bitch', 90),
 ('if', 89),
 ('from', 89),
 ('make', 87),
 ('her', 87),
 ('was', 85),
 ('do', 85)]

In [22]:
rap_chart_common_ctr_2 = Counter()
rap_chart_song_freq_ctr_2 = Counter()
rap_chart_cleaned_all_2 = get_cleaned_album_all(all_rap_chart_lyrics, rap_chart_common_ctr_2, rap_chart_song_freq_ctr_2, 2)

In [23]:
rap_chart_common_ctr_2.most_common(50)

[('in the', 419),
 ('yeah yeah', 359),
 ('i got', 256),
 ('with the', 236),
 ('i dont', 231),
 ('on the', 211),
 ('like a', 206),
 ('on my', 183),
 ('you know', 166),
 ('in my', 155),
 ('and i', 154),
 ('i know', 151),
 ('i aint', 144),
 ('to the', 134),
 ('a nigga', 129),
 ('that i', 127),
 ('it like', 124),
 ('like i', 124),
 ('got a', 123),
 ('walk it', 122),
 ('im a', 117),
 ('but i', 116),
 ('on a', 115),
 ('with me', 113),
 ('i just', 109),
 ('for me', 108),
 ('yeah i', 102),
 ('ayy ayy', 102),
 ('i cant', 95),
 ('in a', 94),
 ('talk it', 92),
 ('a lot', 90),
 ('i like', 89),
 ('i was', 89),
 ('it up', 88),
 ('i talk', 88),
 ('when i', 87),
 ('got the', 86),
 ('it walk', 84),
 ('i want', 82),
 ('pull up', 80),
 ('it i', 80),
 ('for the', 78),
 ('you want', 78),
 ('i can', 76),
 ('i get', 75),
 ('out the', 74),
 ('me and', 73),
 ('you down', 72),
 ('on me', 71)]

In [24]:
rap_chart_song_freq_ctr_2.most_common(50)

[('in the', 113),
 ('i got', 90),
 ('on the', 90),
 ('i dont', 88),
 ('on my', 74),
 ('with the', 71),
 ('you know', 69),
 ('to the', 68),
 ('like a', 66),
 ('in my', 64),
 ('and i', 61),
 ('got a', 61),
 ('i aint', 55),
 ('yeah i', 54),
 ('but i', 53),
 ('i just', 52),
 ('yeah yeah', 51),
 ('that i', 51),
 ('in a', 50),
 ('a nigga', 50),
 ('i cant', 49),
 ('i was', 49),
 ('when i', 48),
 ('i know', 48),
 ('im a', 47),
 ('on a', 45),
 ('with a', 43),
 ('me and', 43),
 ('got the', 41),
 ('with me', 40),
 ('for the', 40),
 ('if you', 39),
 ('i can', 38),
 ('it i', 38),
 ('me i', 37),
 ('from the', 37),
 ('if i', 36),
 ('on me', 36),
 ('pull up', 36),
 ('and im', 34),
 ('up in', 34),
 ('i get', 34),
 ('and you', 33),
 ('all the', 33),
 ('when you', 32),
 ('cause i', 32),
 ('these niggas', 32),
 ('i had', 32),
 ('this shit', 31),
 ('its a', 30)]

In [25]:
inter_rap_gram_2 = rap_chart_common_ctr_2 & rap_chart_song_freq_ctr_2
inter_rap_gram_2.most_common(50)

[('in the', 113),
 ('i got', 90),
 ('on the', 90),
 ('i dont', 88),
 ('on my', 74),
 ('with the', 71),
 ('you know', 69),
 ('to the', 68),
 ('like a', 66),
 ('in my', 64),
 ('and i', 61),
 ('got a', 61),
 ('i aint', 55),
 ('yeah i', 54),
 ('but i', 53),
 ('i just', 52),
 ('yeah yeah', 51),
 ('that i', 51),
 ('in a', 50),
 ('a nigga', 50),
 ('i cant', 49),
 ('i was', 49),
 ('when i', 48),
 ('i know', 48),
 ('im a', 47),
 ('on a', 45),
 ('with a', 43),
 ('me and', 43),
 ('got the', 41),
 ('with me', 40),
 ('for the', 40),
 ('if you', 39),
 ('i can', 38),
 ('it i', 38),
 ('me i', 37),
 ('from the', 37),
 ('on me', 36),
 ('pull up', 36),
 ('if i', 36),
 ('and im', 34),
 ('up in', 34),
 ('i get', 34),
 ('and you', 33),
 ('all the', 33),
 ('when you', 32),
 ('cause i', 32),
 ('these niggas', 32),
 ('i had', 32),
 ('this shit', 31),
 ('its a', 30)]

In [26]:
rap_chart_common_ctr_3 = Counter()
rap_chart_song_freq_ctr_3 = Counter()
rap_chart_cleaned_all_3 = get_cleaned_album_all(all_rap_chart_lyrics, rap_chart_common_ctr_3, rap_chart_song_freq_ctr_3, 3)

In [27]:
rap_chart_common_ctr_3.most_common(50)

[('yeah yeah yeah', 201),
 ('it like i', 84),
 ('it walk it', 84),
 ('like i talk', 83),
 ('walk it like', 82),
 ('i talk it', 82),
 ('ayy ayy ayy', 65),
 ('hot hot hot', 51),
 ('let you down', 44),
 ('gucci gang gucci', 44),
 ('gang gucci gang', 44),
 ('talk it walk', 44),
 ('you know i', 43),
 ('on my mind', 43),
 ('walk it walk', 40),
 ('a lot how', 40),
 ('lot how many', 39),
 ('loco go loco', 37),
 ('bounce that ass', 35),
 ('yeah on god', 33),
 ('woulda gave you', 32),
 ('go loco go', 32),
 ('i like it', 31),
 ('murder on my', 30),
 ('i said i', 28),
 ('i got a', 28),
 ('i got murder', 28),
 ('got murder on', 28),
 ('up in the', 27),
 ('a lot of', 26),
 ('said i like', 26),
 ('oh oh oh', 26),
 ('thats my type', 26),
 ('i got the', 25),
 ('yeah yeah i', 25),
 ('in my bank', 25),
 ('my bank account', 25),
 ('down i break', 25),
 ('i break it', 25),
 ('with the 44', 25),
 ('no no no', 24),
 ('she wanna fuck', 24),
 ('i dont care', 24),
 ('break it down', 24),
 ('walk em down', 24),


In [28]:
rap_chart_song_freq_ctr_3.most_common(50)

[('yeah yeah yeah', 33),
 ('i got a', 27),
 ('up in the', 19),
 ('i had to', 19),
 ('yeah yeah i', 17),
 ('i got the', 16),
 ('im in the', 15),
 ('i know that', 14),
 ('a lot of', 13),
 ('you know i', 13),
 ('i dont know', 13),
 ('i pull up', 13),
 ('i dont want', 12),
 ('i dont wanna', 11),
 ('pull up in', 11),
 ('i need a', 10),
 ('i dont even', 10),
 ('on the beat', 10),
 ('in the back', 10),
 ('all of my', 10),
 ('i dont care', 10),
 ('you know that', 10),
 ('a bad bitch', 10),
 ('i know you', 9),
 ('now i got', 9),
 ('and i dont', 9),
 ('you say you', 9),
 ('you love me', 8),
 ('i used to', 8),
 ('pull up on', 8),
 ('me and my', 8),
 ('dont fuck with', 8),
 ('dont know what', 8),
 ('on the road', 8),
 ('i love you', 8),
 ('im in a', 8),
 ('dont need no', 8),
 ('got a lot', 8),
 ('if you want', 8),
 ('and i know', 8),
 ('oh oh oh', 8),
 ('and you know', 7),
 ('in the club', 7),
 ('you know im', 7),
 ('all in my', 7),
 ('from the back', 7),
 ('up on the', 7),
 ('she got a', 7),
 ('i

In [29]:
inter_rap_gram_3 = rap_chart_common_ctr_3 & rap_chart_song_freq_ctr_3
inter_rap_gram_3.most_common(50)

[('yeah yeah yeah', 33),
 ('i got a', 27),
 ('up in the', 19),
 ('i had to', 19),
 ('yeah yeah i', 17),
 ('i got the', 16),
 ('im in the', 15),
 ('i know that', 14),
 ('a lot of', 13),
 ('you know i', 13),
 ('i dont know', 13),
 ('i pull up', 13),
 ('i dont want', 12),
 ('i dont wanna', 11),
 ('pull up in', 11),
 ('i dont even', 10),
 ('i need a', 10),
 ('on the beat', 10),
 ('in the back', 10),
 ('all of my', 10),
 ('i dont care', 10),
 ('you know that', 10),
 ('a bad bitch', 10),
 ('i know you', 9),
 ('now i got', 9),
 ('and i dont', 9),
 ('you say you', 9),
 ('you love me', 8),
 ('pull up on', 8),
 ('i used to', 8),
 ('me and my', 8),
 ('dont fuck with', 8),
 ('dont know what', 8),
 ('on the road', 8),
 ('i love you', 8),
 ('im in a', 8),
 ('dont need no', 8),
 ('got a lot', 8),
 ('if you want', 8),
 ('and i know', 8),
 ('oh oh oh', 8),
 ('and you know', 7),
 ('all in my', 7),
 ('you know im', 7),
 ('in the club', 7),
 ('from the back', 7),
 ('up on the', 7),
 ('i feel like', 7),
 (

### Intersection Between Quadeca and The Billboard Top Rap Music

In [75]:
inter_1 = inter_ctr_gram_1 & inter_rap_gram_1
inter_2 = inter_ctr_gram_2 & inter_rap_gram_2
inter_3 = inter_ctr_gram_3 & inter_rap_gram_3

In [76]:
inter_1.most_common(50)

[('i', 30),
 ('the', 30),
 ('to', 30),
 ('and', 30),
 ('a', 29),
 ('you', 29),
 ('it', 29),
 ('of', 28),
 ('me', 28),
 ('in', 27),
 ('all', 27),
 ('my', 26),
 ('im', 26),
 ('just', 26),
 ('yeah', 26),
 ('on', 26),
 ('when', 26),
 ('be', 25),
 ('that', 25),
 ('but', 25),
 ('for', 25),
 ('if', 24),
 ('like', 24),
 ('been', 24),
 ('its', 24),
 ('never', 24),
 ('this', 23),
 ('they', 23),
 ('so', 23),
 ('with', 23),
 ('no', 22),
 ('up', 22),
 ('your', 22),
 ('dont', 22),
 ('know', 21),
 ('got', 21),
 ('at', 21),
 ('thats', 20),
 ('do', 20),
 ('shit', 20),
 ('uh', 20),
 ('not', 20),
 ('now', 20),
 ('was', 20),
 ('even', 19),
 ('say', 19),
 ('want', 19),
 ('cant', 19),
 ('get', 19),
 ('still', 19)]

In [77]:
inter_2.most_common(50)

[('in the', 25),
 ('in my', 18),
 ('i got', 17),
 ('on my', 17),
 ('i just', 17),
 ('but i', 15),
 ('yeah i', 15),
 ('when i', 15),
 ('on the', 15),
 ('that i', 14),
 ('of the', 14),
 ('i dont', 13),
 ('im a', 13),
 ('i cant', 13),
 ('in a', 12),
 ('to the', 12),
 ('me i', 12),
 ('i was', 12),
 ('feel like', 12),
 ('yeah yeah', 12),
 ('i been', 12),
 ('so i', 12),
 ('what i', 11),
 ('i need', 11),
 ('if you', 11),
 ('to me', 11),
 ('you know', 11),
 ('i know', 10),
 ('got a', 10),
 ('if i', 10),
 ('this shit', 10),
 ('for real', 10),
 ('i think', 10),
 ('like a', 10),
 ('all of', 10),
 ('and i', 10),
 ('that shit', 10),
 ('when im', 10),
 ('for me', 10),
 ('ive been', 10),
 ('i do', 9),
 ('be the', 9),
 ('all the', 9),
 ('like im', 9),
 ('to be', 9),
 ('but you', 9),
 ('i am', 9),
 ('with the', 9),
 ('it all', 9),
 ('im not', 9)]

In [78]:
inter_3.most_common(50)

[('i got a', 7),
 ('yeah yeah yeah', 7),
 ('i just wanna', 5),
 ('i love you', 5),
 ('up in the', 5),
 ('i had to', 4),
 ('i think i', 4),
 ('i need to', 4),
 ('i dont know', 4),
 ('all of my', 4),
 ('i feel like', 4),
 ('i know i', 3),
 ('wanna be the', 3),
 ('for real i', 3),
 ('i wish i', 3),
 ('feel like im', 3),
 ('i dont wanna', 3),
 ('all of the', 3),
 ('i had a', 3),
 ('me but i', 3),
 ('i was in', 3),
 ('in the back', 3),
 ('the back of', 3),
 ('i dont want', 3),
 ('i just want', 3),
 ('yeah yeah i', 3),
 ('thats why i', 3),
 ('but i cant', 3),
 ('i know that', 3),
 ('it in a', 3),
 ('ooh ooh i', 3),
 ('i know you', 3),
 ('a lot of', 3),
 ('know i got', 2),
 ('just wanna be', 2),
 ('just pulled up', 2),
 ('that i could', 2),
 ('yeah i was', 2),
 ('i do not', 2),
 ('i make a', 2),
 ('yeah i just', 2),
 ('on the way', 2),
 ('think i need', 2),
 ('got so many', 2),
 ('yeah i got', 2),
 ('want me to', 2),
 ('you got a', 2),
 ('when i was', 2),
 ('i did it', 2),
 ('i am not', 2)]

### Differences Between the Two Corpuses

In [81]:
len(set(after_cleaned_all))

2171

In [42]:
quad_not_ms_rap = set(after_cleaned_all) - set(rap_chart_cleaned_all)
len(quad_not_ms_rap)

832

In [85]:
# percent unique
len(quad_not_ms_rap) / len(set(after_cleaned_all))

0.38323353293413176

In [48]:
quad_not_ms_rap

{'$580',
 '*sounds',
 '2010',
 '3:15',
 '45k',
 '911',
 'aah',
 'absolutely',
 'abyss',
 'accepted',
 'acceptin',
 'accustomed',
 'acknowledge',
 'acknowledgin',
 'acne',
 'acquaintance',
 'acted',
 'actions',
 'adds',
 'adjust',
 'adjustin',
 'adlibbers',
 'advanced',
 'advances',
 'aesthetic',
 'afternoon',
 'ages',
 'ago',
 'aiming',
 'aimless',
 'allcaps',
 'along',
 'altitude',
 'alto',
 'ambient',
 'american',
 'amountin',
 'anan',
 'angus',
 'anomaly',
 'antivaxxer',
 'antsy',
 'anxieties',
 'anyways',
 'apparent',
 'archers',
 'arena',
 'arent',
 'arrangement',
 'arrows',
 'art',
 'articulate',
 'artifice',
 'ascent',
 'aside',
 'astray',
 'audibly',
 'august',
 'autopilot',
 'avalanches',
 'backbackbackbackback',
 'backlogged',
 'badtempered',
 'barack',
 'bawl',
 'beaten',
 'begin',
 'believe”',
 'belittle',
 'betray',
 'beyond',
 'bickerin',
 'bipolar',
 'bitch”',
 'blahblah',
 'blastin',
 'blink',
 'bliss',
 'blissful',
 'blisterin',
 'blizzards',
 'bluetooth',
 'blurrin',


In [57]:
ms_rap_not_quad = set(rap_chart_cleaned_all) - set(after_cleaned_all)
len(ms_rap_not_quad)

4630

In [58]:
ms_rap_not_quad

{'troy',
 'sword',
 'minneapolis',
 'guets',
 'nick',
 'serious',
 'precious',
 'purple',
 'poles',
 'powerglide',
 'minaj',
 '21',
 'turns',
 'naked',
 'holding',
 'perico',
 '—instead',
 'fumes',
 'bluff',
 'stacked',
 'tires',
 'gorilla',
 'retreat',
 'exceptions',
 'city',
 'wraith',
 'it—my',
 'chops',
 'wifey',
 'glocky',
 'cheeto',
 'wavy',
 'loud',
 'detroit',
 'blocks',
 'blockbusters',
 'bae',
 'rico',
 'hangman',
 'backbone',
 'happenin',
 'ending',
 'angela',
 'prince',
 'shawtys',
 'kenneled',
 'hundos',
 'quote',
 'networth',
 'lambs',
 'jets',
 'salmon',
 'states',
 'group',
 'bloodsucker',
 'bluray',
 'speaking',
 'dues',
 'kong',
 'whorehouse',
 'richard',
 'lapped',
 'ps',
 'snitching',
 'buried',
 'gun',
 '16yrold',
 'depressing',
 'lobster',
 'joe',
 'duffle',
 'rules',
 'lori',
 'loadin',
 'zoom',
 'deal',
 'hahahahaha',
 'morenita',
 'motorola',
 'candy',
 'willis',
 'rickey',
 'shark',
 'breathin',
 'yep',
 'swing',
 'thizz',
 'chick',
 'scrimmage',
 '10',
 'temp

In [80]:
len(set(after_cleaned_all_2))

8539

In [59]:
quad_not_ms_rap_2 = set(after_cleaned_all_2) - set(rap_chart_cleaned_all_2)
len(quad_not_ms_rap_2)

5815

In [84]:
# percent unique
len(quad_not_ms_rap_2) / len(set(after_cleaned_all_2))

0.6809930905258227

In [50]:
quad_not_ms_rap_2

{'im provin',
 'say okay',
 'said beware',
 'work yeah',
 'a gold',
 'youre livin',
 'a traveller',
 'roll ive',
 'true to',
 'shitty songs',
 'uhuh not',
 'for business',
 'ohoh oh',
 'there wont',
 'tall member',
 'your arms',
 'makin all',
 'down youll',
 'no it',
 'streams than',
 'chatter no',
 'exist if',
 'hater like',
 'going mainstream',
 'until it',
 'plate if',
 'me clearclear',
 'new post',
 'reminds me',
 'and depressed',
 'by saw',
 'uhhuh car',
 'mindnumbin to',
 'teeth gold',
 'the function',
 'curry and',
 'mood they',
 'could forget',
 'like theyre',
 'ayy turn',
 'remind me',
 'surroundings im',
 'twenty racks',
 'afternoon maybe',
 'happy im',
 'her grandkids',
 'balance that',
 'timesd my',
 'shit maybe',
 'why we',
 'my shows',
 'my class',
 'ooh way',
 'im busy',
 'boom bad',
 'im sendin',
 'time but',
 'happy i',
 'expectations rise',
 'footsteps remainеd',
 'mends in',
 'so now',
 'too happy',
 'look where',
 'climb ive',
 'they finish',
 'mean nothing',
 'seat

In [56]:
ms_rap_not_quad_2 = set(rap_chart_cleaned_all_2) - set(after_cleaned_all_2)
len(ms_rap_not_quad_2)

28961

In [53]:
ms_rap_not_quad_2

{'scotts outside',
 'pussy give',
 'cocked it',
 'shut they',
 'be fucked',
 'choke i',
 'land in',
 '2 chainz',
 'hundreds and',
 'many bullets',
 'the street',
 'sing on',
 'a dime',
 'they dead',
 'ghost drive',
 'it bust',
 'hesitatin bullets',
 'be sober',
 'their fucking',
 'mind we',
 'side the',
 'toes that',
 'fucked since',
 'to play',
 'know hard',
 'no extortion',
 'want and',
 'buy balmains',
 'backtoback its',
 'whats his',
 'check on',
 'crazy fuckin',
 'hook it',
 'pay yes',
 'still slangin',
 'moon we',
 'incinerate a',
 'the strap',
 'had really',
 'a fifth',
 'od percs',
 'and nobody',
 '25 hundred',
 'meanie thought',
 'yeah lock',
 'que retumbe',
 'boys childish',
 'that can',
 'tool yeah',
 'kites yeah',
 'count while',
 'store store',
 'swizz beat',
 'really the',
 'the ig',
 'just here',
 'that vitamin',
 'dont wear',
 'watchin i',
 'us is',
 'band ayy',
 'ill beat',
 'issues im',
 'want she',
 'vibes look',
 'graduate i',
 'askin how',
 'real soft',
 'emotions 

In [79]:
len(set(after_cleaned_all_3))

11909

In [62]:
quad_not_ms_rap_3 = set(after_cleaned_all_3) - set(rap_chart_cleaned_all_3)
len(quad_not_ms_rap_3)

10882

In [83]:
# percent unique
len(quad_not_ms_rap_3) / len(set(after_cleaned_all_3))

0.9137627004786296

In [61]:
quad_not_ms_rap_3

{'life looking out',
 'when theres nowhere',
 'you quietly i',
 'got everybody to',
 'being famous they',
 'oh where did',
 'bones apart no',
 'of macaroni now',
 'cd for free',
 'misfit numbers get',
 'start to fumble',
 'music is medicine',
 'they was plotting',
 'and let me',
 'consumed by their',
 'it went the',
 'still beats despite',
 'that is substantial',
 'me like goodbye',
 'ask me what',
 'it takes i',
 'in a blissful',
 'you” i dont',
 'im not fittin',
 'this white noise',
 'the climb youre',
 'so if im',
 'my last oh',
 'huh b lou',
 'with that feelin',
 'bout me facts',
 'lookin up to',
 'cant remember please',
 'fall but a',
 'slowly kill himself',
 'a lotta new',
 'yeah and its',
 'i will have',
 'she think she',
 'hear a lot',
 'ashes i dont',
 'know which makes',
 'bitch back it',
 'away been like',
 'left fuck a',
 'rap facts you',
 'lil mellow mood',
 'i said— i',
 'listen i want',
 'so stop askin',
 'easy to see',
 'awkward i acted',
 'pay me run',
 'while the cloc

In [65]:
ms_rap_not_quad_3 = set(rap_chart_cleaned_all_3) - set(after_cleaned_all_3)
len(ms_rap_not_quad_3)

49856

In [66]:
ms_rap_not_quad_3

{'gang smackin high',
 'in never lost',
 'you seen it',
 'thirsty and in',
 'fly yeah xans',
 'ill marry her',
 'mind my pimpin',
 'hearts i guess',
 'good better put',
 'me crazy i',
 'me go the',
 'forward to the',
 'alyx shit nike',
 'the love that',
 'had never played',
 'facts i keep',
 'hit my eses',
 'i love my',
 'bitch her ass',
 'the ball gimme',
 'these streets alright',
 'and not with',
 'me but when',
 'dont care ill',
 'get near boujee',
 'rather you trust',
 'facts facts i',
 'for what i',
 'on another nigga',
 'passed us they',
 'drink til im',
 'two three i',
 'like every time',
 'you last night',
 'all bad my',
 'love to do',
 'clique im moonwalkin',
 'its sweet wanna',
 'it all the',
 'all night all',
 'serious girl yeah',
 'drop hey ayy',
 'shit they wanted',
 'it up til',
 'i free em',
 'my dad told',
 'taste chitown you',
 'talking new woo',
 'to keep peace',
 'bonnie and clyde',
 'gang gucci gang',
 'you know shit',
 'go wheres the',
 'im sick no',
 'could just s

### Overall Observations Between the Two Corpuses

Here are some of the notable 1-grams between both Quadeca's corpus and the Billboard Rap Music Corpus:
> Quadeca: `i, you, me, all, my, im, your, dont, know`
> Billboard Rap Music: `i, me, you, my, im, dont, like`

>It seems that the most common 1-grams between the two corpuses are very similar. They seem to be focusing on a 'me versus you' type of theme. I found that the word `i` was used in nearly every song from Quadeca's corpus and was used in every song in the Billboard Rap Music corpus. I think this was expected because rappers tend to talk about themselves a lot in their songs and they tend to compare themselves to other rappers.

Here are some of the notable 2-grams between both Quadeca's corpus and the Billboard Rap Music Corpus:
> Quadeca: `in the, in my, i got, on my, i just, yeah i, when i, that i, i think, i dont, im a, i cant, and i, i am`
> Billboard Rap Music: `in the, i got, i dont, on my, and i, got a, i aint, yeah i, in a, i was, when i, if i`

>Again, I noticed the usage of the term `i` within the 2-grams. Once again, the 2-grams between these two corpuses are very similar. A lot of the 2-grams are referring to the rapper and what they're thinking of themselves and what materialistic wealth they have. This can be seen in the 2-grams: `i think, i am, im a, i got`

Here are some of the notable 3-grams between both Quadeca's corpus and the Billboard Rap Music Corpus:
> Quadeca: `i got a, i think i, i need to, i just wanna, i do not, i love you, im on my, i had to, just pulled up, i dont know`
> Billboard Rap Music: `yeah yeah yeah, i got a, i got the, i know that, i dont know, pull up in, i dont know`

>Once again, the 3-grams between these two corpuses are similar and they both tend to talk about the rapper themselves. In both corpuses we see the rapper talking about having something over some else (`i got a, i got the`) or arriving in a what I assume would be a fancy car (`just pulled up, pull up in`). In Quadeca's corpus, we see that there are 3-grams that refer to feeling and love in the upper top 25, whereas in the Billboard Rap Music corpus, we don't see 3-grams that refer to feelings and love until the bottom 25 3-grams.

Other things that I've noticed where that even though the n-grams were fairly the same across the board, there are still differences in the n-grams that Quadeca uses and that the Billboard Rap Music uses. In fact, there are 832 unique 1-grams that are used by Quadeca that aren't used by Billboard Rap Music artist. When it comes to 2-grams, there are 5,815 unique bigrams that are used by Quadeca that aren't used by Billboard Rap Music artist. Finally, for the 3-grams, there are 10,882 unique trigrams that are used by Quadeca that aren't used by Billboard Rap Music artist.

### Conclusion

After looking through the intersections of the intersections of the Counters, as well as the set difference between the two corpuses, I found that even though Quadeca's most common 1-/2-/3-grams are very similar to the Billboard Rap Music 1-/2-/3-grams, there are still quite a lot of unique tokens, bigrams, and trigrams that Quadeca uses that aren't found in the Billboard Rap Music corpus. In the case of 1-grams, Quadeca has 38.32% unique tokens. In the case of 2-grams, Quadeca has 68.10% unique bigrams. In the case of 3-grams, Quadeca has 91.38% unique trigrams. Although the 1-grams have a fairly low unique percentage, there are only so many words you can use that make sense in a rap. However, with the bigrams and trigrams having 68.10% uniqueness and 91.38% uniqueness respectively, I think it is safe to say that Quadeca's lyrics are fairly different from the lyrics of the Billboard Rap Music. 

In [86]:
%store after_all_lyrics
%store all_rap_chart_lyrics

Stored 'after_all_lyrics' (list)
Stored 'all_rap_chart_lyrics' (list)
